python/dgl/transform.py

```diff
@@ -143,6 +146,345 @@ def segmented_knn_graph(x, k, segs):
+
+    The shared edge features are the concatenation of the edge features
+    from g1 and g2.
+
+    >>> ug.edata['h']
+    tensor([[3.],
+            [4.],
+            [5.],
+            [6.],
+            [7.],
+            [8.]])
+
+    **In-place operations on features of FIRST graph will be reflected
+    on features of the joint_unioned graph. Out-place operations will not
+    be reflected.**
+    >>> ug.ndata['h'] += 1
+    >>> g.ndata['h']
+    tensor([[1.],
+            [2.],
+            [3.],
+            [1.]])
+    >>> g.ndata['h'] += 1
+    >>> ug.ndata['h']
+    tensor([[2.],
+            [3.],
+            [4.],
+            [2.]])
+    >>> ug.ndata['h2'] = th.ones(4, 1)
+    >>> 'h2' in g.ndata
+    False
+    >>> g.edata['h'] += 1
+    >>> ug.edata['h']
+    tensor([[3.],
+            [4.],
+            [5.],
+            [6.],
+            [7.],
+            [8.]])
+
+    **Heterographs with Multiple Edge Types**
+    >>> g1 = dgl.heterograph({
+    >>>     ('user', 'follows', 'user'): (th.tensor([0, 2]), th.tensor([1, 2])),
+    >>>     ('user', 'plays', 'game'): (th.tensor([1, 2, 1]), th.tensor([2, 1, 1]))
+    >>> })
+    >>> g1.nodes['game'].data['hv'] = th.ones(3, 1)
+    >>> g1.edges['plays'].data['he'] = th.zeros(3, 1)
+    >>> g2 = dgl.heterograph({
+    >>>     ('user', 'follows', 'user'): (th.tensor([2]), th.tensor([1])),
+    >>>     ('user', 'plays', 'game'): (th.tensor([0, 0, 1]), th.tensor([2, 1, 0]))
+    >>> })
+    >>> g2.edges['plays'].data['he'] = th.zeros(3, 1)
+    >>> g3 = dgl.heterograph({
+    >>>     ('user', 'follows', 'user'): (th.tensor([1]), th.tensor([2])),
+    >>>     ('user', 'plays', 'game'): (th.tensor([1]), th.tensor([0]))
+    >>> })
+    >>> g3.edges['plays'].data['he'] = th.zeros(1, 1)
+
+    The joint_union of the above three heterographs can be obtained by
+    combining joint_union of the subgraphs corresponding to
+    ('user', 'follows', 'user') and the subgraphs corresponding to
+    ('user', 'plays', 'game').
+
+    >>> ug = dgl.joint_union([g1, g2, g3])
+    >>> ug
+    Graph(num_nodes={'game': 3, 'user': 3},
+          num_edges={('user', 'follows', 'user'): 4, ('user', 'plays', 'game'): 7},
+          metagraph=[('user', 'user'), ('user', 'game')])
+    >>> ug.edges(etype='follows')
+    (tensor([0, 2, 2, 1]), tensor([1, 2, 1, 2]))
+    >>> ug.edges(etype='plays')
+    (tensor([1, 2, 1, 0, 0, 1, 1]), tensor([2, 1, 1, 2, 1, 0, 0]))
+    >>> ug.nodes['game'].data
+    {}
+    >>> ug.edges['plays'].data
+    {}
+    """
+    assert len(graph_list) > 1, 'joint_union requires at least two graphs as input'
+    # check all graphs have the same canonical etypes
+    canonical_etypes_0 = graph_list[0].canonical_etypes
+    for i in range(1, len(graph_list)):
+        canonical_etypes_i = graph_list[i].canonical_etypes
+        assert len(canonical_etypes_0) == len(canonical_etypes_i), \
+            'graph_{} should has the same canonical edge type as graph_0'.format(i)
```

Comment:
1. "graph_{}" -> "{}-th graph"
2. "has" -> "have"
3. "graph_0" -> "the first graph"